src/mount-support.c

```diff
@@ -276,14 +277,47 @@ void setup_snappy_os_mounts()
 	// moved all paths to /tmp/snap.rootfs_*. Because we are using unshare with
 	// CLONE_NEWNS we can essentially use pivot_root just like chroot but this
 	// makes apparmor unaware of the old root so everything works okay.
+
+	// The code below was taken fomr src/lxc/conf.c in github.com/lxc/lxc
+	// It is available under the terms of the LGPLv2.1 and higher and
+	// was written by various LXC developers.
+
 	debug("chrooting into %s", rootfs_dir);
-	if (chdir(rootfs_dir) == -1) {
-		die("cannot change working directory to %s", rootfs_dir);
+
+	int oldroot = -1, newroot = -1;
```

Comment:
Yeah and I think we very much want to be in control as to when we close those fds at the end as that's when the kernel will actually flush the old mount tree.

As for the error paths, well, we're about to exit, so closing the fds is a bit of a waste of time.